Quiz List SwiftUI

barbarakos opened the pull request:
Created screen with list of quizzes divided in categories.
![Simulator Screen Shot - iPhone 14 Plus - 2022-12-09 at 12 57 13](https://user-images.githubusercontent.com/101199634/206699098-4a9d4519-ea36-4caf-935a-9d860424bb06.png)

barbarakos committed:
Quiz List SwiftUI

Created screen with list of quizzes divided in categories.

barbarakos committed:
Added title to QuizListView

barbarakos committed:
Merge branch 'feature/UserSwiftUI' into feature/QuizListSwiftUI

# Conflicts:
#	QuizApp.xcodeproj/project.pbxproj

barbarakos committed:
Updated project.pbxproj

barbarakos committed:
SwiftLint warnings

barbarakos committed:
Added navigation bar title.

barbarakos committed:
Edit background

barbarakos committed:
Edit UserView background

barbarakos committed:
Merge branch 'feature/UserSwiftUI' into feature/QuizListSwiftUI

# Conflicts:
#	QuizApp/User/UserView.swift

barbarakos committed:
Merge branch 'feature/UserSwiftUI' into feature/QuizListSwiftUI

# Conflicts:
#	QuizApp/User/ViewControllers/UserViewController.swift

barbarakos committed:
Deleted QuizViewController

barbarakos committed:
Update AppDependencies.swift

marinzg commented on QuizApp/Extensions/LinearGradient.swift:
```diff
@@ -0,0 +1,14 @@
+                                             Color(red: 0.154, green: 0.185, blue: 0.463)]),
+                              startPoint: .top,
+                              endPoint: .bottom)
+    }
```
I see you've done that here, but please just do sth with line wrapping and formatting.

marinzg commented on QuizApp/Quiz/DifficultyViews.swift:
```diff
@@ -0,0 +1,55 @@
+            view1 = .yellow
+            view2 = .yellow
+            view3 = .yellow
+        }
```
Maybe to avoid this switch/case, you could do sth like
```swift
// first
.foregroundColor(.yellow)
// second
.foregroundColor(type != .easy ? .yellow : emptyFillColor)
// third
.foregroundColor(type == .hard ? .yellow : emptyFillColor)
```

Because you'll always have to check, either with switch/case, if/else, ternary, but this way you're explicitly saying which view has which color in which case.
The naming `viewX: Color` is a bit confusing as well.

marinzg commented on QuizApp/Quiz/Models/CategorySection.swift:
```diff
@@ -20,4 +21,17 @@ enum CategorySection: String, CaseIterable {
         }
     }
 
+    var colorSwiftUI: Color {
```
Why SwiftUI suffix?

marinzg commented on QuizApp/Quiz/QuizCellView.swift:
```diff
@@ -0,0 +1,70 @@
+
+struct QuizCellView: View {
+
+    private var quiz: QuizModel!
```
Why implicit unwrap?

marinzg commented on QuizApp/Quiz/QuizListView.swift:
```diff
@@ -0,0 +1,96 @@
+        }
+    }
+
+    func setSegmentedControlAppearance() {
```
Can this be `private`?

marinzg commented on QuizApp/Quiz/QuizListView.swift:
```diff
@@ -0,0 +1,96 @@
+    }
+
+    func getQuizzes() {
+        let allCategories = CategorySection.allCases.map { $0.rawValue }
```
`segmentationalSelection` can also be an `@Published` var in view model to which you can subscribe there (combine sink) and do whatever you need and then update some other published var which will update the view.

marinzg commented on QuizApp/Quiz/QuizListView.swift:
```diff
@@ -0,0 +1,96 @@
+        ScrollView {
+            VStack(alignment: .leading, spacing: 20) {
+                Picker("", selection: $segmentationSelection) {
+                    let items = ["All"] + CategorySection.allCases.map {$0.rawValue}
```
This should be in VM.

marinzg commented on QuizApp/Quiz/QuizListView.swift:
```diff
@@ -0,0 +1,96 @@
+                        Text(option)
+                    }
+                }
+                .onChange(of: segmentationSelection) { _ in
+                    getQuizzes()
+                }
```
You can get rid of this if you move `segmentedSelection` into view model.

marinzg commented on QuizApp/Quiz/QuizListView.swift:
```diff
@@ -0,0 +1,96 @@
+                .pickerStyle(SegmentedPickerStyle())
+                .cornerRadius(10)
+
+                let allCategories = CategorySection.allCases.map { $0.rawValue }
+                if allCategories.contains(segmentationSelection) {
+                    let section = CategorySection(rawValue: segmentationSelection)!
+                    Section(header: Text(section.rawValue)
+                        .sectionHeaderStyle(section)) {
+                            ForEach(viewModel.quizzes, id: \.self) { quiz in
+                                QuizCellView(quiz: quiz)
+                                    .onTapGesture {
+                                        viewModel.showQuizDetails(quiz: quiz)
+                                    }
+                            }
+                            .cornerRadius(30)
+                        }
+                } else {
+                    ForEach(CategorySection.allCases, id: \.self) { section in
+                        let filteredQuizzes = viewModel
+                            .quizzes
+                            .filter { $0.category == section.rawValue.uppercased() }
+                        if !filteredQuizzes.isEmpty {
+                            Section(header: Text(section.rawValue)
+                                .sectionHeaderStyle(section)) {
+                                    ForEach(filteredQuizzes, id: \.self) { quiz in
+                                        QuizCellView(quiz: quiz)
+                                            .onTapGesture {
+                                                viewModel.showQuizDetails(quiz: quiz)
+                                            }
+                                    }
+                                    .cornerRadius(30)
+                                }
+                        }
+                    }
+                }
```
Try to move as much as possible to view model and I think you'll be able to have only one `Section` view which will be updated and filtered through ViewModel.

barbarakos committed:
Merge branch 'feature/UserSwiftUI' into feature/QuizListSwiftUI

# Conflicts:
#	QuizApp.xcodeproj/project.pbxproj

barbarakos committed:
PR fix

I had to change one of the computed properties' name in CategorySection to colorUIKit.
It will be deleted once I delete the SearchViewController.

barbarakos committed:
add weak self inside main dispatch queue

barbarakos committed:
Filtering quizzes by section in view model

dorastipkovic commented on QuizApp/Extensions/Color+Extensions.swift:
```diff
@@ -0,0 +1,21 @@
+    }
+
+    static var incorrect: Color {
+        return Color(red: 0.988, green: 0.395, blue: 0.395)
```
You can omit `return`

barbarakos committed:
Changed base URL

barbarakos committed:
Update Color+Extensions.swift
